### PR TITLE
GraphQL mutations refactoring: input types, cart mutations & multi-cart preparation

### PR DESCRIPTION
**Pull request by pozylon**
This PR contains massive name refactorings / api surface changes.

## Removed (BREAKING):

* captureCart (has never been implemented)

## Renamed (BREAKING):

* checkout to checkoutCart
* input type FilterQuery to FilterQueryInput
* input type ProductConfigurationParameter to ProductConfigurationParameterInput
* input type HashedPassword to HashedPasswordInput
* updateOrder to updateCart
* Order.address to Order.billingAddress
* Parameter address to billingAddress in updateCart

## New order manipulating mutation parameters in preparation of "multi-cart":

* updateCart: new optional orderId
* checkoutCart: new optional orderId
* addCartProduct: new optional orderId
* addCartDiscount: new optional orderId

## New convenience mutations:

* emptyCart(orderId: ID)

That way, a very simple checkout is possible without fetching the orderId once in one big query but the full feature set is only available when using the indirect methods.

### CODE REVIEW

**Comment by macrozone** [comment]
i would be cautious with confusing orders with carts. That it is the same is just a technical detail of your implementation, but for a user, its fundamentally different. I can't just update the quantity of an order after it has been payed. so `rename updateCartItemQuantity to updateOrderItemQuantity` makes no sense to me. It should be `updateCartItemQuantity`

so i would suggest:

- updateCart* mutatons stays the same, but additionally take an optional cartId for multi carts. if no cartId is given, the default one is used
- emptyCart removes everything in a cart
- checkoutOrder should be checkoutCart. It transforms the cart to an order (and returns the order)

Also concerning mutations that actually should be able to update an order (because of backoffice-jobs):

- make it clear what they do and for who their are intended for. `updateOrder` for example is very generic. It should not be easy to just update stuff on an order outside of the normal flow. Maybe even prefix these functions:

`mutations.manage.updateOrder`

or also for product manage functions

`mutations.manage.updateProduct`

The api should focus first on a developer who implements the shop frontend and second on a developer implementing backoffice apps. Altough unchained has a good acl implementation, its hard to reason about what can be done by whom, just looking at the graphql api

File: packages/api/resolvers/mutations/updateOrderItemQuantity.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+export default function (root, { itemId, quantity }, { userId }) {
+  log(`mutation updateOrderItemQuantity ${itemId} ${quantity}`, { userId });
+  if (quantity === 0) throw new OrderQuantityTooLowError({ data: { quantity } });
+  const item = OrderPositions.findOne({ _id: itemId });
+  if (!item) throw new OrderItemNotFoundError({ data: { itemId } });
+  const order = item.order();
+  if (order.status !== OrderStatus.OPEN) {
+    throw new OrderWrongStatusError({ data: { status: order.status } });
+  }
```
Comment (macrozone) [inline packages/api/resolvers/mutations/updateOrderItemQuantity.js]: also cart here

**Comment by pozylon** [comment]
So let's review all the order/cart related mutations today:

addCartProduct
addCartDiscount
updateCart
checkoutCart

addOrderProduct
updateOrderItemQuantity
removeOrderItem
addOrderDiscount
removeOrderDiscount
updateOrder
removeOrder
checkoutOrder
confirmOrder
payOrder
setOrderPaymentProvider
setOrderDeliveryProvider
updateOrderDeliveryShipping
updateOrderDeliveryPickUp
updateOrderPaymentCard
updateOrderPaymentInvoice
updateOrderPaymentPostfinance
updateOrderPaymentPaypal
updateOrderPaymentCrypto

Today, only confirmOrder, payOrder and removeOrder are actual methods that should be allowed when the order is not a cart anymore. Following your argumentation, we would stick with Cart as the name for all the other mutations and take an optional cartId for all mutations above that currently require an orderId, resulting in:

addCartProduct(cartId: ID, ...)
addCartDiscount(cartId: ID, ...)
updateCart(cartId: ID, ...)
checkoutCart(cartId: ID, ...)
updateCartItemQuantity
removeCartItem
removeCartDiscount
setCartPaymentProvider
setCartDeliveryProvider
updateCartDeliveryShipping
updateCartDeliveryPickUp
updateCartPaymentCard
updateCartPaymentInvoice
updateCartPaymentPostfinance
updateCartPaymentPaypal
updateCartPaymentCrypto

removeOrder
confirmOrder
payOrder


prefixing with mutations.manage is not a good idea due to https://github.com/graphql/graphql-js/issues/221#issuecomment-157481861

**Comment by pozylon** [comment]
updateOrder is definitely missleading in the new naming

**Comment by pozylon** [comment]
The more I think about it I guess we should definetly make the user know that a cart is an open order. But the updateCart namings are much more elegant.

What about having something like:
addCartProduct(orderId: ID, ...)

I mean the GraphQL type that gets returned is an Order too, having that link is not only a technical implementation detail but also an important knowhow of the whole concept. Then they know that cart's _id stays persistent throughout the checkout for example. And I think it could be pretty straightforward to document that lifecycle.

**Comment by pozylon** [comment]
So I would go for this set of mutations:

addCartProduct(orderId: ID, ...)
addCartDiscount(orderId: ID, ...)
updateCart(orderId: ID, ...)
checkoutCart(orderId: ID, ...)
updateCartItemQuantity
removeCartItem
removeCartDiscount
setCartPaymentProvider(orderId: ID, ...)
setCartDeliveryProvider(orderId: ID, ...)
updateCartDeliveryShipping
updateCartDeliveryPickUp
updateCartPaymentCard
updateCartPaymentInvoice
updateCartPaymentPostfinance
updateCartPaymentPaypal
updateCartPaymentCrypto

removeOrder(orderId: ID!, ...)
confirmOrder(orderId: ID!, ...)
payOrder(orderId: ID!, ...)

**Comment by macrozone** [comment]
> prefixing with mutations.manage is not a good idea due to [graphql/graphql-js#221 (comment)](https://github.com/graphql/graphql-js/issues/221#issuecomment-157481861)

yeah. Ok so maybe have a clear description on these methods?

File: packages/api/resolvers/mutations/updateOrderItemQuantity.js (hunk 1 above)
Comment (pozylon) [reply packages/api/resolvers/mutations/updateOrderItemQuantity.js]: updated, please see

File: packages/api/resolvers/mutations/removeOrderItem.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import { OrderItemNotFoundError, OrderWrongStatusError } from '../../errors';
+
+export default function (root, { itemId }, { userId }) {
+  log(`mutation removeOrderItem ${itemId}`, { userId });
+  const orderItem = OrderPositions.findOne({ _id: itemId });
+  if (!orderItem) throw new OrderItemNotFoundError({ data: { orderItem } });
+  const order = orderItem.order();
+  if (order.status !== OrderStatus.OPEN) {
+    throw new OrderWrongStatusError({ data: { status: order.status } });
+  }
```
[the remark this answers is not in the document]
Comment (pozylon) [reply packages/api/resolvers/mutations/removeOrderItem.js]: updated, please see

File: packages/api/resolvers/mutations/removeOrderDiscount.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import { OrderDiscountNotFoundError, OrderWrongStatusError } from '../../errors';
+
+export default function (root, { discountId }, { userId }) {
+  log(`mutation removeOrderDiscount ${discountId}`, { userId });
+  const orderDiscount = OrderDiscounts.findOne({ _id: discountId });
+  if (!orderDiscount) throw new OrderDiscountNotFoundError({ data: { orderDiscount } });
+  const order = orderDiscount.order();
+  if (order.status !== OrderStatus.OPEN) {
+    throw new OrderWrongStatusError({ data: { status: order.status } });
+  }
```
[the remark this answers is not in the document]
Comment (pozylon) [reply packages/api/resolvers/mutations/removeOrderDiscount.js]: The OrderWrongStatus is an API error (ApolloError)

[the remark this answers is not in the document]
Comment (pozylon) [reply packages/api/resolvers/mutations/removeOrderDiscount.js]: Good idea with the encapsulation of orderstatus.open

File: packages/api/resolvers/mutations/updateCart.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  if (meta) {
+    order = order.updateContext(meta);
+  }
+  if (billingAddress) {
+    order = order.updateBillingAddress({ ...billingAddress, countryCode: countryContext });
+  }
+  if (contact) {
+    order = order.updateContact({ contact });
+  }
+  return order;
```
[the remark this answers is not in the document]
Comment (pozylon) [reply packages/api/resolvers/mutations/updateCart.js]: Me too, maybe we can look at this later, not feeling that a refactoring could lift's it's weight here

**Comment by pozylon** [comment]
Thanks @macrozone for the review, please see the updated issue comment about the final schema changes.

**Pull request merged**